courier/handler.go

```diff
@@ -119,30 +125,82 @@ func (h *Handler) listCourierMessages(w http.ResponseWriter, r *http.Request, _
 		}
 	}
 
-	x.PaginationHeader(w, urlx.AppendPaths(h.r.Config().SelfAdminURL(r.Context()), AdminRouteMessages), int64(tc), filter.Page, filter.PerPage)
+	w.Header().Set("x-total-count", fmt.Sprint(tc))
+	keysetpagination.Header(w, r.URL, nextPage)
 	h.r.Writer().Write(w, r, l)
 }
 
-func parseMessagesFilter(r *http.Request) (ListCourierMessagesParameters, error) {
+func parseMessagesFilter(r *http.Request) (ListCourierMessagesParameters, []keysetpagination.Option, error) {
 	var status *MessageStatus
 
 	if r.URL.Query().Has("status") {
 		ms, err := ToMessageStatus(r.URL.Query().Get("status"))
 
 		if err != nil {
-			return ListCourierMessagesParameters{}, err
+			return ListCourierMessagesParameters{}, nil, err
 		}
 
 		status = &ms
 	}
 
-	page, itemsPerPage := x.ParsePagination(r)
+	opts, err := keysetpagination.Parse(r.URL.Query(), keysetpagination.NewMapPageToken)
+	if err != nil {
+		return ListCourierMessagesParameters{}, nil, err
+	}
+
 	return ListCourierMessagesParameters{
-		RequestParameters: migrationpagination.RequestParameters{
-			Page:    page,
-			PerPage: itemsPerPage,
-		},
 		Status:    status,
 		Recipient: r.URL.Query().Get("recipient"),
-	}, nil
+	}, opts, nil
+}
+
+// Get Courier Message Parameters
+//
+// swagger:parameters getCourierMessage
+// nolint:deadcode,unused
+type getCourierMessage struct {
+	// MessageID is the ID of the message.
+	//
+	// required: true
+	// in: path
+	MessageID string `json:"id"`
+}
+
+// swagger:route GET /admin/courier/messages/{id} courier getCourierMessage
+//
+// # Get a Message
+//
+// Gets a specific messages by the given ID.
+//
+//	Produces:
+//	- application/json
+//
+//	Security:
+//		oryAccessToken:
+//
+//	Schemes: http, https
+//
+//	Responses:
+//		200: message
+//		400: errorGeneric
+//		default: errorGeneric
+func (h *Handler) getCourierMessage(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	msgID, err := uuid.FromString(ps.ByName("msgID"))
+
+	if err != nil {
+		h.r.Writer().WriteError(w, r, herodot.ErrBadRequest.WithError(err.Error()).WithDebugf("could not parse parameter {id} as UUID, got %s", ps.ByName("id")))
+		return
+	}
+
+	message, err := h.r.CourierPersister().FetchMessage(r.Context(), msgID)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	if !h.r.Config().IsInsecureDevMode(r.Context()) {
+		message.Body = "<redacted-unless-dev-mode>"
```

Comment:
@jonas-jonas can you create a follow up issue for this? I don't know where we discussed a possible solution, but it would be great to add the idea we had there.